framework/src/engine/consensus/consensus.ts

```diff
@@ -514,14 +518,19 @@ export class Consensus {
 					block,
 				});
 
-				this._chain.validateBlock(block, {
-					version: BLOCK_VERSION,
-				});
-				const previousLastBlock = objects.cloneDeep(lastBlock);
+				try {
+					this._chain.validateBlock(block, {
+						version: BLOCK_VERSION,
+					});
+				} catch (error) {
+					throw new ApplyPenaltyError(error);
+				}
+
 				await this._deleteBlock(lastBlock);
 				try {
 					await this._executeValidated(block);
 				} catch (err) {
+					const previousLastBlock = objects.cloneDeep(lastBlock);
```

Comment:
`cloneDeep` was meant to be executed before deleting the block to avoid possibility of mutating the last block. (lastBlock is reference)